docs/sphinx/developers/cpp/tutorial.txt

```diff
@@ -0,0 +1,406 @@
+:cpp:func:`getImageCount` method is used to find the number of images.
+This is then used to safely loop through each of the available images.
+Each of the :cpp:func:`getPixelsSizeA` methods takes the image index
+as its only argument.  Internally, this is used to find the
+:cpp:class:`Image` model object for the specified index, and then call
+the :cpp:func:`getSizeA` method on that object and return the result.
+Since objects can contain other objects, some accessor methods require
+the use of more than one index.  For example, an :cpp:class:`Image`
+object can contain multiple :cpp:class:`Plane` objects.  Similar to
+the above example, there is a :cpp:func:`getPlaneCount` method,
+however since it is contained by an :cpp:class:`Image` it has an
+additional image index argument to get the plane count for the
+specified image.  Likewise its accessors such as
+:cpp:func:`getPlaneTheZ` take two arguments, the image index and
+the plane index.  Internally, these indices will be used to find the
+:cpp:class:`Image`, then the :cpp:class:`Plane`, and then call
+:cpp:func:`getTheZ`.  When using the :cpp:class:`MetadataRetrieve`
+interface with an :cpp:class:`OMEXMLMetadata` store, the methods are
+simply a shorthand for navigating through the tree of model objects.
+
+.. literalinclude:: examples/metadata-io.cpp
+   :language: cpp
+   :start-after: query-example-start
+   :end-before: query-example-end
+
+The methods for storing data using the :cpp:class:`MetadataStore`
+interface are similar.  The set methods use the same indices as the
+get methods, with the value to set as an additional initial argument.
+The following example demonstrates how to update dimension sizes for
+images in the store:
+
+.. literalinclude:: examples/metadata-io.cpp
+   :language: cpp
+   :start-after: update-example-start
+   :end-before: update-example-end
+
+When adding new objects to the store, as opposed to updating existing
+ones, some additional considerations apply.  An new object is added to
+the store if the object corresponding to an index does not exist and
+the index is the current object count (i.e. one past the end of the
+last valid index).  Note that for data model objects with a
+:cpp:func:`setID` method, this method alone will trigger insertion and
+must be called first, before any other methods which modify the
+object.  The following example demonstrates the addition of a new
+:cpp:class:`Image` to the store, plus contained :cpp:class:`Plane`
+objects.
+
+.. literalinclude:: examples/metadata-io.cpp
+   :language: cpp
+   :start-after: add-example-start
+   :end-before: add-example-end
+
+Full example source: :download:`metadata-io.cpp <examples/metadata-io.cpp>`
+
+.. seealso::
+
+  - :doxygen:`Metadata classes <namespaceome_1_1xml_1_1meta.html>`
+  - :doxygen:`createID <namespaceome_1_1bioformats.html#ab3bf80ec03bcf20b199ce2761d48fe01>`
+  - :doxygen:`createOMEXMLMetadata <namespaceome_1_1bioformats.html#a094cce3a9581c222d128cc9f366aaa75>`
```

Comment:
That's a good question.  I'm not certain; I hoped that the hash was stable--hash of the namespace/class/method signature.  If it's not, this will need rethinking.

According to http://sourceforge.net/p/doxygen/mailman/message/30539258/ it's a hash of the filename + line number which isn't so great.  If this can be configured better, that would be good.  I'll take a look.